src/execution/operator/helper/physical_buffered_batch_collector.cpp

```diff
@@ -0,0 +1,120 @@
+// Sink
+//===--------------------------------------------------------------------===//
+class BufferedBatchCollectorGlobalState : public GlobalSinkState {
+public:
+	mutex glock;
+	weak_ptr<ClientContext> context;
+	shared_ptr<BufferedData> buffered_data;
+};
+
+BufferedBatchCollectorLocalState::BufferedBatchCollectorLocalState() {
+}
+
+SinkResultType PhysicalBufferedBatchCollector::Sink(ExecutionContext &context, DataChunk &chunk,
+                                                    OperatorSinkInput &input) const {
+	auto &gstate = input.global_state.Cast<BufferedBatchCollectorGlobalState>();
+	auto &lstate = input.local_state.Cast<BufferedBatchCollectorLocalState>();
+
+	lock_guard<mutex> l(gstate.glock);
+	lstate.current_batch = lstate.BatchIndex();
+
+	auto batch = lstate.BatchIndex();
+	auto min_batch_index = lstate.GetMinimumBatchIndex();
+	auto &buffered_data = dynamic_cast<BatchedBufferedData &>(*gstate.buffered_data);
+	buffered_data.UpdateMinBatchIndex(min_batch_index);
+
+	if (!lstate.blocked || buffered_data.ShouldBlockBatch(batch)) {
+		lstate.blocked = true;
+		auto callback_state = input.interrupt_state;
+		auto blocked_sink = BlockedSink(callback_state, chunk.size());
+		buffered_data.BlockSink(blocked_sink, batch);
+		return SinkResultType::BLOCKED;
+	}
+
+	auto to_append = make_uniq<DataChunk>();
+	to_append->Initialize(Allocator::DefaultAllocator(), chunk.GetTypes());
+	chunk.Copy(*to_append, 0);
+
+	buffered_data.Append(std::move(to_append), batch);
+	return SinkResultType::NEED_MORE_INPUT;
+}
+
+SinkNextBatchType PhysicalBufferedBatchCollector::NextBatch(ExecutionContext &context,
+                                                            OperatorSinkNextBatchInput &input) const {
+
+	auto &gstate = input.global_state.Cast<BufferedBatchCollectorGlobalState>();
+	auto &lstate = input.local_state.Cast<BufferedBatchCollectorLocalState>();
+
+	lock_guard<mutex> l(gstate.glock);
+
+	auto batch = lstate.current_batch;
+	auto min_batch_index = lstate.GetMinimumBatchIndex();
+	auto new_index = lstate.BatchIndex();
+
+	auto &buffered_data = dynamic_cast<BatchedBufferedData &>(*gstate.buffered_data);
+	buffered_data.CompleteBatch(batch);
+	lstate.current_batch = new_index;
+	// FIXME: this can move from 'other' chunks to 'current' chunks, increasing the 'current_batch_tuple_count'
+	// We might want to block here if 'current_batch_tuple_count' has already reached the threshold
+	// So we don't completely disregard the BUFFER_SIZE we set
+	buffered_data.UpdateMinBatchIndex(min_batch_index);
+	return SinkNextBatchType::READY;
+}
+
+SinkCombineResultType PhysicalBufferedBatchCollector::Combine(ExecutionContext &context,
+                                                              OperatorSinkCombineInput &input) const {
+	auto &gstate = input.global_state.Cast<BufferedBatchCollectorGlobalState>();
+	auto &lstate = input.local_state.Cast<BufferedBatchCollectorLocalState>();
+
+	lock_guard<mutex> l(gstate.glock);
+
```

Comment:
This can probably be removed, if NextBatch is called, and no Sink is called inbetween with the new batch index, we wont have any batches to flush anyways

Reply:
Actually, seems to be necessary still...